### DISCUSSION TITLE
Option to save <ChatGPT> buffers by default

### DISCUSSION THREAD

**Discussion by moritzschaefer** [karthink/gptel#129] [Ideas]
It would be nice to have the configurable option to automatically associated the gptel buffers (named <ChatGPT #n> by default with a file (e.g. in a configurable directory with file name <date>.txt). This way, without extra effort, one can maintain a well-accessible history of chat interactions.

If not within `gptel`, do you have a recommendation how to implement this in the easiest manner?

**Comment by karthink** [comment]
There are many ways to do it:

- Save the gptel buffer to a file as needed.  I take it this is the step you want to automate.

- Save the buffer to a file when `gptel-mode` is turned on: (Note: untested)
```emacs-lisp
(defun my/gptel-write-buffer ()
  "Save buffer to disk when starting gptel"
  (unless (buffer-file-name (current-buffer))
    (let ((suffix (format-time-string "%Y%m%dT%H%M" (current-time)))
          (chat-dir "~/chatgpt-log"))
      (unless (file-directory-p chat-dir)
        (make-directory chat-dir :parents))
      (write-file (expand-file-name (concat "gptel-" suffix ".txt") chat-dir)))))

(add-hook 'gptel-mode-hook #'my/gptel-write-buffer)
```
This will save new gptel sessions to `~/chatgpt-log` with names like `gptel-20231105T1026.txt`.  Or you can save to `default-directory`, or to `(project-root (project-current))` to save them at the root of your current project in Emacs.

- I use a [single file per project](https://github.com/karthink/.emacs.d/blob/master/init.el#L3601) (always named `Chat.org`) to keep persistent logs of ChatGPT conversations, and bind summoning it to the project keymap (`C-x p`), where all the other project commands are. Chat topics are separated by Org headers with `gptel-set-topic`.

This is a feature best tailored by users for their individual workflows, and thus not something I plan to implement in gptel -- so I'm moving this thread to the [discussions](https://github.com/karthink/gptel/discussions) section.

**Reply by agzam** [reply]
Does it really matter what file extension you'd use? Correct me if I'm wrong, I think the package currently handles markdown and txt better than org (proly because LLMs don't handle org-mode very well), is it better to set the extension according `gptel-default-mode`?

**Reply by karthink** [reply]
I don't think the extension matters, so long as it's one of txt, md or org, since these are the ones supported by `gptel-mode`.

> I think the package currently handles markdown and txt better than org

It should handle Org fine, I think.  gptel converts from markdown to org as required.

**Reply by agzam** [reply]
> It should handle Org fine

Oh, very nice, I'm gonna change `gptel-default-mode` to org. I might have confused it with `chatgpt-shell`. I don't know though how good is its support of org-mode these days, I stopped using it a while ago.

**Reply by agzam** [reply]
Btw, it looks like you may want to update the docstring for the var, it currently says:

```
gptel-default-mode
The default major mode for dedicated chat buffers.

If markdown-mode is available, it is used.  Otherwise gptel defaults to text-mode.
```

I tried setting it to org-mode and it works nicely.

**Comment by agzam** [comment]
I have another question about this approach. 

It looks like that it sets some local vars in the header of the buffer, inserting e.g., something like:

```
<!-- Local Variables: -->
<!-- gptel-model: "claude-3-5-sonnet-20240620" -->
<!-- gptel--backend-name: "Claude" -->
<!-- gptel--bounds: nil -->
<!-- End: -->
```

That's nice, but then it doesn't do anything when you switch the model. I wonder if the mode-hook is the best one for this task?

**Reply by karthink** [reply]
> That's nice, but then it doesn't do anything when you switch the model.

Not sure how that matters.  It records the model and other details when you save the buffer.

**Reply by agzam** [reply]
What I'm saying is that it doesn't change those vars afterwards, even if you switch to a different model. Not a big deal. I think this can be fixed by additionally hooking to another hook.

**Reply by karthink** [reply]
I tried the following:

1. Start a chat (or turn on `gptel-mode`)
2. Save the buffer. This writes the backend and model details to the file.
4. Now change the model, system message or other backend details.
5. Save the file again.  This **updates** the model details in the file.

So I'm not sure what you mean by this:
> it doesn't change those vars afterwards, even if you switch to a different model

So it's either a bug or a misunderstanding.

**Reply by agzam** [reply]
> This updates the model details in the file.

Yeah, I expected exactly that, alas that didn't do it for some reason. Maybe something is messing it up in my setup, I'll try to find out. Also, I think switching gptel-default-mode to org sometimes breaks it in a weird way where the api stops responding, I hit that twice, I'll keep an eye on that as well.

**Reply by karthink** [reply]
Please create an issue (or two) if you can find a way to reproduce these bugs.
